### ISSUE TITLE
freebsd compile assistance

### ISSUE THREAD

**Issue by jungle-boogie** [Expensify/Bedrock#63]
Hi All,

Thanks for creating and maintaining this project!

What's happening here when using gmake on freebsd?

```
g++ -g -std=gnu++11 -DSVERSION="\"c178f76262d25959e73b2ecb82e5925f23045b40\"" -Wall -I/usr/home/sean/bin/Bedrock -I/usr/home/sean/bin/Bedrock/mbedtls/include -MMD -MF libstuff/libstuff.d -MT libstuff/libstuff.h.gch -c libstuff/libstuff.h
/bin/sh: g++: not found
gmake: *** No rule to make target 'libstuff/libstuff.d', needed by '.build/libstuff/SData.d'.  Stop.
```

```
% gmake -v
GNU Make 4.2.1
Built for i386-portbld-freebsd11.0
Copyright (C) 1988-2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by amidvidy** [comment]
try `GXX=clang++ CC=clang gmake`

**Comment by jungle-boogie** [comment]
Gets a little further...

```
% GXX=clang++ CC=clang gmake
clang++ -g -std=gnu++11 -DSVERSION="\"c178f76262d25959e73b2ecb82e5925f23045b40\"" -Wall -I/usr/home/sean/bin/Bedrock -I/usr/home/sean/bin/Bedrock/mbedtls/include -MMD -MF libstuff/libstuff.d -MT libstuff/libstuff.h.gch -c libstuff/libstuff.h
clang++: warning: treating 'c-header' input as 'c++-header' when in C++ mode, this behavior is deprecated
libstuff/libstuff.h:4:9: warning: #pragma once in main file [-Wpragma-once-outside-header]
#pragma once
        ^
libstuff/libstuff.h:15:10: fatal error: 'pcrecpp.h' file not found
#include <pcrecpp.h> // sudo apt-get install libpcre++-dev
         ^
1 warning and 1 error generated.
gmake: *** No rule to make target 'libstuff/libstuff.d', needed by '.build/libstuff/SData.d'.  Stop.
```

**Comment by tylerkaraszewski** [comment]
Hey there, we're making two changes that you might find helpful.

1) We're removing the dependency on pcre (see here: https://github.com/Expensify/Bedrock/pull/62)

2) We've disabled the pre-compiled headers for OS X, because thus far we haven't tried to make them work with clang. See here: https://github.com/Expensify/Bedrock/blob/master/Makefile#L45

You might need to do something similar for freebsd (I'm not sure what string `uname -s` returns on freebsd, but you might just substitute it for "Darwin" if you have trouble here).

**Comment by jungle-boogie** [comment]
@tylerkaraszewski,

I did see the pull request for removing pcre yesterday or so. After that's merged, I'll attempt a build again.

My uname is simply FreeBSD. After 1 is completed, I'll update the makefile and substitute with freebsd to see what happens.

Thanks!

**Comment by tylerkaraszewski** [comment]
You can also just checkout that branch and see if it works for you, if you don't want to wait. :)

**Comment by jungle-boogie** [comment]
I see you reverted the patch after it was in master for a little bit:
https://github.com/Expensify/Bedrock/commit/2cb8c4029725a572543b3e10003a0cac678a6f42

With this being the case, can you add code to point to the pcrecpp.h?

**Comment by quinthar** [comment]
@tylerkaraszewski What's the latest plan here?

**Comment by tylerkaraszewski** [comment]
We currently have no plan here. Do we want to add freebsd to a list of supported platforms?

**Comment by quinthar** [comment]
Oh, not that Expensify would support, but I'm open to it being community
supported.  I'm more wondering what your plan is for the regexp library --
my recollection is you went to std::regex, but it was like 10x slower and
thus reverted?  I ask because I think it was the pcre library that broke
the freebsd build, so knowing your plan with regex would affect the
community's plan for compiling on freebsd.

**Comment by tylerkaraszewski** [comment]
I don't think we have any plans to switch back to std:regex at this point, the performance was just too poor.

**Comment by quinthar** [comment]
Reminds me of the Boost::Spirit library.  So insanely powerful, so insanely
slow.  K, thanks!

**Comment by Vadius** [comment]
FreeBSD has PCRE support, but if it installed as port (to `/usr/local` as mentioned above), clang will not find it by default (see https://wiki.freebsd.org/WarnerLosh/UsrLocal).
I think here is 2 options how to solve issue:
* make GCC the only supported compiler for FreeBSD
* modify Makefile for better support of clang on FreeBSD (or explicitly use /usr/local for -I and -L options)

**Comment by jungle-boogie** [comment]
Well if option 1 is the default, will that still require the commit that was backed out?

**Comment by Vadius** [comment]
@jungle-boogie , I've tried to build the code on FreeBSD 11.0, but there is a couple of code issues which make build impossible, e.g.: https://github.com/Expensify/Bedrock/blob/master/libstuff/SSignal.cpp#L5 misses case of FreeBSD OS.
So, I think, there should be a set of code changes to build the project on BSD platform.

PS. Tried to compile with `clang` because `gcc` crashes on my fresh system (maybe my fault, didn't use FreeBSD since v4.4)

**Comment by jungle-boogie** [comment]
> So, I think, there should be a set of code changes to build the project on BSD platform.

Ah, good point. Once the code is added, I'm happy to test it out.
